File: scripts/arrow-ipc-oracle.ts
import {createHash} from 'node:crypto';
import {backend} from '../native/arrow/runtime';
import {importArrowSchema,exportArrowSchemaIpc,proposeArrowNodeEdit} from '../src';
const cases=await Bun.file('fixtures/arrow/schema-cases.json').json();const results=[];
for(const c of cases){
 const doc=importArrowSchema(JSON.stringify(c.input),{id:c.id});
 try{const bytes=exportArrowSchemaIpc(doc,backend);const file=c.id+'.umf.arrow';await Bun.write('fixtures/arrow/'+file,bytes);results.push({id:c.id,status:'exported',file,sha256:createHash('sha256').update(bytes).digest('hex')});}
 catch(e){results.push({id:c.id,status:'blocked',message:(e as Error).message});}
}
const blocked=results.filter(r=>r.status==='blocked').map(r=>r.id);
if(blocked.join(',')!=='type-47,type-48,type-49,duplicate-metadata'||results.length!==52)throw Error('Unexpected public IPC baseline');
const doc=importArrowSchema(JSON.stringify(cases[1].input),{id:'edit'});const edit=proposeArrowNodeEdit(doc,'/fields/0/name','"edited_field"').document;
await Bun.write('fixtures/arrow/edited.umf.arrow',exportArrowSchemaIpc(edit,backend));
await Bun.write('fixtures/arrow/public-ipc-results.json',JSON.stringify({backend:backend.identity,cases:52,exported:48,blocked:4,results},null,2)+'\n');console.log({exported:48,blocked:4,edited:true});

File: scripts/arrow-ipc-oracle.py
import hashlib,json
from pathlib import Path
import pyarrow as pa
base=Path('fixtures/arrow');report=json.loads((base/'public-ipc-results.json').read_text())
assert report['exported']==48 and report['blocked']==4
count=0
for row in report['results']:
    if row['status']=='blocked':continue
    data=(base/row['file']).read_bytes();assert hashlib.sha256(data).hexdigest()==row['sha256']
    with pa.ipc.open_stream(data) as actual,pa.ipc.open_stream((base/(row['id']+'.arrow')).read_bytes()) as expected:
        assert actual.schema.equals(expected.schema,check_metadata=True),row['id']
        assert actual.read_all().num_rows==0
    count+=1
with pa.ipc.open_stream((base/'edited.umf.arrow').read_bytes()) as edited:
    assert edited.schema.names==['edited_field']
    assert edited.schema.field(0).type==pa.bool_()
assert count==48
(base/'public-ipc-oracle-results.json').write_text(json.dumps({'oracle':'pyarrow '+pa.__version__,'schemas':count,'metadataAndNativeTypeEquality':True,'editedFieldVerified':True,'scope':'Schema-only IPC, compared with the pinned native baseline. No record-batch, dictionary-value, unknown FlatBuffer or arbitrary stream import claim.'},indent=2)+'\n')
print('Arrow public IPC: 48 native schemas and edited field verified independently')
